ci: move twister to be pull_request base, integrate compliance

nashif opened the pull request:
move twister workflow to be pyll_request target and make it depend on compliance check passing. If compliance check fails, no twister will be run until the compliance check has passed.

In cases where check_compliance workflow has a fall positivie, a new label can be used to override and let the workflow continue. The label to be used for override is `Compliance: False Positive`

Fixes https://github.com/zephyrproject-rtos/infrastructure/issues/21

nashif committed:
ci: twister: convert to pull_request

Convert twister workflow to use pull_request.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ci: compliance: convert compliance to workflow_call

Add compliance to twister workflow as a the first step. If compliance
fials, no twister will run.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

pdgendt commented on .github/workflows/twister.yaml:
```diff
@@ -20,7 +20,11 @@ concurrency:
+    uses: ./.github/workflows/compliance.yml
+
   twister-build-prep:
+    needs: compliance-check
```
Does this mean compliance needs to succeed before twister will run?

pdgendt commented:
Is this normal?

![image](https://github.com/user-attachments/assets/12f690d4-1336-47ed-b8b0-445f227989e5)

fabiobaltieri commented:
> Is this normal?
> 
> ![image](https://private-user-images.githubusercontent.com/3170234/392027147-12f690d4-1336-47ed-b8b0-445f227989e5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyNDkxMjAsIm5iZiI6MTczMzI0ODgyMCwicGF0aCI6Ii8zMTcwMjM0LzM5MjAyNzE0Ny0xMmY2OTBkNC0xMzM2LTQ3ZWQtYjhiMC00NDVmMjI3OTg5ZTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMDNUMTgwMDIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2U1NDcyMDRlMTYzYTgzZDFiNzVhZjc1OTY2ZmVmZTI2NjQ0NzYxNzcyYjI4MzZmYjA4NzdmMzRkZTY4MGQ2YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.dWJtmicZq4KL159Zfj9kRhlb9mbNcvyLhxB2FZT2jAM)

it's not, give it a force push.

nashif commented:
> Is this normal?

yes, the compliance workflow is gone now, the status of the twister workflow (we probably need to rename this to something else now) will be used for PRs, so we need to remove this from GH settings.

nashif commented:
btw, this was testing here:

https://github.com/zephyrproject-rtos/zephyr-testing/actions/runs/12096547199?pr=301

fabiobaltieri commented:
> yes, the compliance workflow is gone now, the status of the twister workflow (we probably need to rename this to something else now) will be used for PRs, so we need to remove this from GH settings.

Oh so it's not going to show as its own workflow anymore, but then it's not going to retrigger on PR message edits right?

fabiobaltieri commented on .github/workflows/compliance.yml:
```diff
@@ -1,15 +1,20 @@
 on:
-  pull_request:
-    types:
-    - edited
```
I think this is going to break the "empty PR message" below, the check would fail and only rerun on push. It should probably be moved on a separate workflow.

nashif replied:
yeah i know and I am aware of that. I am thinking of moving this type of checks to somewhere else.

nashif replied:
@fabiobaltieri  should be "fixed" now

nashif committed:
ci: move PR body check to a different workflow

Move PR body check to a workflow that monitors editing of PRs.
Avoids rerunnin compliance check and other heavy weights workflows on
just an edit of the PR body.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ci: workflows: rename do_not_merge workflow

Add more functionality to the workflow, it is not only about not-merging
anymore.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
huh, it actually works here in the same PR, I did not expect that. Nice


https://github.com/zephyrproject-rtos/zephyr/actions/runs/12149516128/job/33880348191?pr=82457

nashif commented on .github/workflows/twister-publish.yaml:
```diff
@@ -12,7 +12,7 @@ jobs:
   upload-to-elasticsearch:
     if: |
       github.repository == 'zephyrproject-rtos/zephyr' &&
-      github.event.workflow_run.event != 'pull_request_target'
+      github.event.workflow_run.event != 'pull_request'
```
but this never runs on PRs, so what is there to validate?

nashif replied:
@stephanosio  afaik, a workflow_run cant be changed and run in a PR, it needs to be in the tree, and this one does not run on PRs, so check for validity is not needed.

stephanosio replied:
> but this never runs on PRs, so what is there to validate?

Apparently lack of sleep can do wonders, including making `!=` look like `==`.

> afaik, a workflow_run cant be changed and run in a PR

That is correct.

nashif commented on .github/workflows/compliance.yml:
```diff
@@ -1,15 +1,20 @@
+  workflow_call:
 
 jobs:
+  check_compliance_push:
+    if: github.event_name == 'push'
+    runs-on: ubuntu-22.04
+    name: Run compliance checks on patch series (PR)
+    steps:
+    - name: Do nothing
+      run: |
+        echo "Doing nothing"
```
workaround for now, if nothing runs here (on push), dependant PRs will never run. Was not able to find a better way.
Eventually we can also run some compliance checks on main if needed.

stephanosio commented on .github/workflows/compliance.yml:
```diff
@@ -121,10 +120,12 @@ jobs:
+    - name: resolve
+      if: |
+        steps.check-warns.outputs.exit_code != 0 &&
+        ! contains(github.event.pull_request.labels.*.name, 'Compliance: False Positive')
```
Maybe we should make the `PR Metadata Check` workflow fail when a PR is labeled `Compliance: False Positive` so that a manual override is necessary; otherwise, this can be quite easily abused ...

kartben commented on .github/workflows/compliance.yml:
```diff
@@ -1,15 +1,20 @@
 jobs:
+  check_compliance_push:
+    if: github.event_name == 'push'
+    runs-on: ubuntu-22.04
```
is it time we move all gh actions to 24.04?

nashif replied:
yeah, need to do this after that...

nashif committed:
ci: compliance: add compat mode

Temporary workflow to satisfy requirements when dealing with old PRs.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

carlescufi commented on .github/workflows/compliance.yml:
```diff
@@ -1,18 +1,23 @@
-    - opened
-    - reopened
-    - synchronize
+  workflow_call:
```
I understand the motivation to move the twister job from `pull_request_target` to `pull_request`, but what is the motivation behind changing this?

nashif replied:
https://github.com/zephyrproject-rtos/infrastructure/issues/21

carlescufi replied:
Sure, but that is a very wide proposal. This PR implements something partially different, and I was wondering whether we could simplify things by just having the compliance workflow run on `pull_request` and then have the twister workflow be triggered on `workflow_run`. To me this sounds like a simpler solution and equally safe, since the only thing that would run automatically is on `pull_request` with no write access.

nashif replied:
> This PR implements something partially different,

how is this different? The ultimate goal is to have a super-workflow with depenecies and a way to fail fast, which this PR does. see here https://github.com/zephyrproject-rtos/zephyr-testing/actions/runs/13906350118 for how this looks like

This PR uses existing twister workflow to start compliance, you are asking for compliance to start twister. I am sure I have tried this last year and it did not work or something was missing.
There are lots of special cases that are different from the doc workflow, i.e. we might fail compliance but we still want to run twister, twister being skipped because nothing has changed and other things.
I can spend some time going backwards and doing all the steps to show what does not work, which is not fun and will take time.

carlescufi replied:
All I am asking is that we document why we did it in this (in my  opinion) counter-intuitive way. If we are to go back in time later and look at how this happened, for me this PR and its commits does not provide a clear explanation of why. I don't typically block PRs but with the amount time I've spent on GH actions I've learned to document things very clearly for our future selves :)
